lib/encoding.ts

```diff
@@ -78,22 +78,95 @@ function decodeNullableBoolean(value: CompactNullableBoolean): boolean | undefin
     return undefined;
 }
 
-export function encodeRules(rules: AutoConsentCMPRule[]): CompactCMPRuleset {
-    const strings: string[] = [];
+/**
+ * Builds a map of strings to their indices, preserving the indices of existing strings, when possible.
+ * @param existingStrings - The strings array from the previous version of the ruleset.
+ * @param rules - The rules to collect strings from.
+ * @returns A new strings array.
+ */
+function buildStrings(existingStrings: string[], rules: AutoConsentCMPRule[]): string[] {
+    // First pass: collect all strings that will be used
+    const usedStrings = new Set<string>();
 
-    function replaceStrings(selector: string): number {
-        let index = strings.indexOf(selector);
-        if (index === -1) {
-            index = strings.push(selector) - 1;
+    function collectStrings(step: AutoConsentRuleStep) {
+        for (const [longKey] of compactedRuleSteps) {
+            if (step[longKey] && typeof step[longKey] === 'string') {
+                usedStrings.add(step[longKey]);
+            }
+        }
+        if (step.if) {
+            collectStrings(step.if);
+            step.then?.forEach(collectStrings);
+            step.else?.forEach(collectStrings);
+        }
+        if (step.any) {
+            step.any.forEach(collectStrings);
         }
-        return index;
     }
 
+    // Collect all used strings from all rules
+    rules.forEach((rule) => {
+        (rule.prehideSelectors || []).forEach((selector) => usedStrings.add(selector));
+        rule.detectCmp.forEach(collectStrings);
+        rule.detectPopup.forEach(collectStrings);
+        rule.optOut.forEach(collectStrings);
+        (rule.test || []).forEach(collectStrings);
+    });
+
+    // Separate existing strings that are still used vs new strings
+    const existingUsedStrings: Array<{ str: string; originalIndex: number }> = [];
+    const newStrings: string[] = [];
+
+    usedStrings.forEach((str) => {
+        const existingIndex = existingStrings.indexOf(str);
+        if (existingIndex !== -1) {
+            existingUsedStrings.push({ str, originalIndex: existingIndex });
+        } else {
+            newStrings.push(str);
+        }
+    });
+
+    // Create compact array, trying to preserve original indices
```

Comment:
I was struggling to follow this a bit, and I _think_ it's overcomplicated things a bit. I think we just need 2 steps: Copying the still-used values from the previous array over at the same index, then filling everything else in the remaining gaps. I think the following code should work (though I didn't test it), and we can get rid of the block above (`existingUsedStrings` construction) and below.

```suggestion
    // New array of undefined, ready to be filled.
    const strings = new Array(usedStrings.size).fill()
    // copy existing values to the same, capping to the max array length
    existingStrings.slice(0, usedStrings.size).forEach((s, i) => {
      if (usedStrings.has(s)) {
        strings[i] = s;
      }
    })
    // copy remaining strings into empty slots. This will also place any existing strings whose previous indices are out of the bounds of the new array.
    strings.forEach((s) => {
      if (strings.indexOf(s) === -1) {
        const nextFreeSlot = strings.indexOf() // find next undefined value
        strings[nextFreeSlot] = s
      }
    })
    return strings
```